### ISSUE TITLE
out variant of many loss functions are not consistent with non-out variant when reduction is not none

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#50382]
## 🐛 Bug

The out variants of `l1_loss`, `mse_loss`, `smooth_l1_loss`  loss functions implement the reduction with `.mean(dim=0)` while non-out variants of the same loss functions implement the reduction with `.mean()` which should be the correct behavior.

## To Reproduce

```
auto input = torch::randn({5,6});
auto target = torch::randn({5,6});

auto result = torch::empty_like(input);
at::mse_loss_out(result, input, target);
auto output = at::mse_loss(input, target);

ASSERT_EQ(result.sizes(), output.sizes()); // This should fail because result has size {6} and out has size {}
```

## Expected behavior

They should be consistent

## Environment

PyTorch version: 1.8.0a0+3435892
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 8.4.0-3ubuntu2) 8.4.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 3070
Nvidia driver version: 455.38
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] pytorch-sphinx-theme==0.0.19
[pip3] torch==1.8.0a0
[conda] numpy                     1.19.4                   pypi_0    pypi
[conda] pytorch-sphinx-theme      0.0.19                   pypi_0    pypi
[conda] torch                     1.8.0a0                   dev_0    <develop>

**Comment by albanD** [comment]
Thanks for the report!
I don't think this is high pri (and agree with the labelling) as these out variants don't support autograd and computing a loss and not computing gradients is a fairly niche use case.
But we should definitely fix that if we can.

**Comment by gchanan** [comment]
is this also C++ only?  We don't expose out variants of nn functions in python do we?

**Comment by soulitzer** [comment]
> is this also C++ only? We don't expose out variants of nn functions in python do we?
Yep this should be C++ only. Theres `F.normalize`, but that should be the only one that accepts an `out` arg.
